File: ValueSequencer/CHexagramSequences.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Text;

namespace ValueSequencer
{
        public class CHexagramSequences
        {
                public CHexagramValueSequencer AutoCast(ref CHexagramValueSequencer hvs)
                {
                        Random r = Sequences.m_ranSession ?? new Random(DateTime.Now.Millisecond);
                        for (int l = 0; l < 6; ++l)
                        {
                                int count = (r.Next(5) + 1) * 100 + r.Next(100);
                                for (int c = 0; c < count; ++c)
                                        hvs.Trigram(l / 3).Line(l % 3).Next(true);
                        }
                        return hvs;
                }
        }
	public class CHexagram : IComparable
	{
		public CHexagram(ref CHexagramValueSequencer hvsPrimary)
		{
			m_hvsPrimary = new CHexagramValueSequencer(ref hvsPrimary);
		}

		public string DescribeCast
		{
			get { return m_hvsPrimary.DescribeCast(); }
		}

		public int CompareTo(object obj)
		{
			CHexagram h = (CHexagram)obj;
			return m_hvsPrimary.HexagramId().CompareTo(h.m_hvsPrimary.HexagramId());
		}

		public void Add() => ++Count;
		public int Count { get; private set; } = 0;

		private CHexagramValueSequencer m_hvsPrimary;
	}

	public class CHexagramArray //: IEnumerable
	{
		public CHexagramArray()
		{
			CHexagramValueSequencer hvsPrimary = new CHexagramValueSequencer(0);
			hvsPrimary.First();
			for (int p = 0; p < 64; ++p)
			{
				for (int s = 0; s < 64; ++s)
				{
					CHexagramValueSequencer hvs = new CHexagramValueSequencer(ref hvsPrimary);
					for (int l = 0; l < 6; ++l)
						if (((s & (1 << l)) >> l) == 1)
							hvs.Trigram(l / 3).Line(l % 3).Next(false);
					Add(ref hvs);
				}
				hvsPrimary.Next();
			}
			Array.Sort<CHexagram>(m_arrHexagram);
		}

		public void Add(ref CHexagramValueSequencer hvsPrimary)
		{
			m_arrHexagram[++m_nCount] = new CHexagram(ref hvsPrimary);
		}

		public CHexagramArray MultiCast(int nCount)
		{
			CHexagramValueSequencer hvs = new CHexagramValueSequencer(63); // more even ????????
			for (int i = 0; i < nCount; ++i)
			{
				//CHexagramValueSequencer hvs = new CHexagramValueSequencer(63);
				AutoCast(ref hvs);
				CHexagram h = new CHexagram(ref hvs);
				int nIndex = Array.BinarySearch<CHexagram>(m_arrHexagram, h);
				if(nIndex >= 0)
				m_arrHexagram[nIndex].Add();
			}
			return this;
		}

		public CHexagramValueSequencer AutoCast(ref CHexagramValueSequencer hvs) //////// Random ?????????????????????????????
		{
			Random r = true ? Sequences.m_ranSession : new Random(DateTime.Now.Millisecond);
			for (int l = 0; l < 6; ++l)
			{
				int count = (r.Next(5) + 1) * 100 + r.Next(100);
				for (int c = 0; c < count; ++c)
					hvs.Trigram(l / 3).Line(l % 3).Next(true);
			}
			return hvs;
		}

		public IEnumerator GetEnumerator()
		{
			throw new NotImplementedException();
		}

		public CHexagram[] HexagramArray() => m_arrHexagram;
		public CHexagram this[int i] => m_arrHexagram[i];

		private CHexagram[] m_arrHexagram = new CHexagram[4096];
		private int m_nCount = -1;
	}
}

File: ValueSequencer.ts/CHexagramSequences.ts
import { CHexagramValueSequencer } from "./CHexagramValueSequencer";
import { Sequences } from "./Sequences";

export class CHexagramSequences {
  public autoCast(hvs: CHexagramValueSequencer): CHexagramValueSequencer {
    const random = Sequences.randomSession;
    for (let l = 0; l < 6; ++l) {
      const count = (random.next(5) + 1) * 100 + random.next(100);
      for (let c = 0; c < count; ++c) {
        hvs.trigram(Math.floor(l / 3)).line(l % 3).next(true);
      }
    }
    return hvs;
  }
}

export class CHexagram {
  private m_hvsPrimary: CHexagramValueSequencer;
  private _count = 0;

  constructor(hvsPrimary: CHexagramValueSequencer) {
    this.m_hvsPrimary = new CHexagramValueSequencer(hvsPrimary);
  }

  get describeCast(): string {
    return this.m_hvsPrimary.describeCast();
  }

  public compareTo(other: CHexagram): number {
    return this.m_hvsPrimary.hexagramId().localeCompare(other.m_hvsPrimary.hexagramId());
  }

  public add(): void {
    this._count += 1;
  }

  get count(): number {
    return this._count;
  }
}

export class CHexagramArray {
  private m_arrHexagram: CHexagram[] = [];

  constructor() {
    const hvsPrimary = new CHexagramValueSequencer(0);
    hvsPrimary.first();
    for (let p = 0; p < 64; ++p) {
      for (let s = 0; s < 64; ++s) {
        const hvs = new CHexagramValueSequencer(hvsPrimary);
        for (let l = 0; l < 6; ++l) {
          if (((s & (1 << l)) >> l) === 1) {
            hvs.trigram(Math.floor(l / 3)).line(l % 3).next(false);
          }
        }
        this.add(hvs);
      }
      hvsPrimary.next();
    }
    this.m_arrHexagram.sort((a, b) => a.compareTo(b));
  }

  public add(hvsPrimary: CHexagramValueSequencer): void {
    this.m_arrHexagram.push(new CHexagram(hvsPrimary));
  }

  public multiCast(nCount: number): this {
    const hvs = new CHexagramValueSequencer(63);
    for (let i = 0; i < nCount; ++i) {
      this.autoCast(hvs);
      const h = new CHexagram(hvs);
      const index = this.m_arrHexagram.findIndex((item) => item.compareTo(h) === 0);
      if (index >= 0) {
        this.m_arrHexagram[index].add();
      }
    }
    return this;
  }

  public autoCast(hvs: CHexagramValueSequencer): CHexagramValueSequencer {
    const random = Sequences.randomSession;
    for (let l = 0; l < 6; ++l) {
      const count = (random.next(5) + 1) * 100 + random.next(100);
      for (let c = 0; c < count; ++c) {
        hvs.trigram(Math.floor(l / 3)).line(l % 3).next(true);
      }
    }
    return hvs;
  }

  public hexagramArray(): CHexagram[] {
    return this.m_arrHexagram;
  }

  public at(index: number): CHexagram {
    return this.m_arrHexagram[index];
  }
}
